pkg/kubelet/status/status_manager.go

```diff
@@ -850,14 +878,28 @@ func (m *manager) needsUpdate(uid types.UID, status versionedPodStatus) bool {
 	if !ok {
 		return false
 	}
-	return m.canBeDeleted(pod, status.status)
+	return m.canBeDeleted(pod, status.status, status.podIsFinished)
 }
 
-func (m *manager) canBeDeleted(pod *v1.Pod, status v1.PodStatus) bool {
+func (m *manager) canBeDeleted(pod *v1.Pod, status v1.PodStatus, podIsFinished bool) bool {
 	if pod.DeletionTimestamp == nil || kubetypes.IsMirrorPod(pod) {
 		return false
 	}
-	return m.podDeletionSafety.PodResourcesAreReclaimed(pod, status)
+	// Delay deletion of pods until the phase is terminal.
+	if !podutil.IsPodPhaseTerminal(pod.Status.Phase) {
+		klog.V(3).InfoS("Delaying pod deletion as the phase is non-terminal", "phase", status.Phase, "pod", klog.KObj(pod), "podUID", pod.UID)
+		return false
+	}
+	// If this is an update completing pod termination then we know the pod termination is finished.
+	if podIsFinished {
+		klog.V(3).InfoS("The pod termination is finished as SyncTerminatedPod completes its execution", "phase", status.Phase, "pod", klog.KObj(pod), "podUID", pod.UID)
+		return true
+	}
+	if m.podDeletionSafety.PodIsFinished(pod) {
```

Comment:
Can you recreate the scenario where you needed this to still be here?  I.e. did it happen in a test job you can link?

Reply:
It was before I cherry-picked your branch, so might be worth to repeat it, but basically the scenario was that `SyncTerminatedPod` (which sets `podIsFinished`) failed. Then, syncBatch had to lookup the workers. This might not be needed if we ensure that all versions in versionedPodStatus are true following `SyncTerminatedPod`.

Reply:
Ok, yeah I'd prefer removal after my change, which removes one method of pod safety.

Reply:
Removed, testing yet, but it does not seem necessary indeed.

Reply:
Continues to pass in testing. Resolving the thread.